### ISSUE TITLE
Bug: no-warning-comments does not work with multiline comments

### ISSUE THREAD

**Issue by Reeywhaar** [eslint/eslint#16103]
There is exact same issue, that marked as **resolved**, but it seems there is regression?

**Previous issue**: https://github.com/eslint/eslint/issues/9884

I will copy description from linked issue, except I've updated the environment. 

> I'm going to report using the online demo rather than my own project, so that anyone else can reproduce this bug as well. It happens in the online demo. https://eslint.org/demo

> [Direct link to the demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyoqXG4gKiB0b2RvXG4gKi9cblxuLypcbiAqIHRvZG9cbiAqL1xuXG4vLyB0b2RvXG5cbnZhciBmb28gPSBiYXI7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo1LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Miwibm8td2FybmluZy1jb21tZW50cyI6Mn0sImVudiI6e319fQ==)

### Environment

Node version: 18.2.0
npm version: 8.9.0
Local ESLint version: 8.13.0
Global ESLint version: 8.13.0
Operating System: macOS 12.4

### What parser are you using?

@babel/eslint-parser

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "parserOptions": {
        "ecmaVersion": 5,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "constructor-super": 2,
        "no-case-declarations": 2,
        "no-class-assign": 2,
        "no-compare-neg-zero": 2,
        "no-cond-assign": 2,
        "no-console": 2,
        "no-const-assign": 2,
        "no-constant-condition": 2,
        "no-control-regex": 2,
        "no-debugger": 2,
        "no-delete-var": 2,
        "no-dupe-args": 2,
        "no-dupe-class-members": 2,
        "no-dupe-keys": 2,
        "no-duplicate-case": 2,
        "no-empty-character-class": 2,
        "no-empty-pattern": 2,
        "no-empty": 2,
        "no-ex-assign": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-semi": 2,
        "no-fallthrough": 2,
        "no-func-assign": 2,
        "no-global-assign": 2,
        "no-inner-declarations": 2,
        "no-invalid-regexp": 2,
        "no-irregular-whitespace": 2,
        "no-mixed-spaces-and-tabs": 2,
        "no-new-symbol": 2,
        "no-obj-calls": 2,
        "no-octal": 2,
        "no-redeclare": 2,
        "no-regex-spaces": 2,
        "no-self-assign": 2,
        "no-sparse-arrays": 2,
        "no-this-before-super": 2,
        "no-undef": 2,
        "no-unexpected-multiline": 2,
        "no-unreachable": 2,
        "no-unsafe-finally": 2,
        "no-unsafe-negation": 2,
        "no-unused-labels": 2,
        "no-unused-vars": 2,
        "no-useless-escape": 2,
        "require-yield": 2,
        "use-isnan": 2,
        "valid-typeof": 2,
        "no-warning-comments": 2
    },
    "env": {}
}
```
</details>

```js
/**
 * todo
 */

/*
 * todo
 */

// todo

var foo = bar;
```

Run it in the [online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyoqXG4gKiB0b2RvXG4gKi9cblxuLypcbiAqIHRvZG9cbiAqL1xuXG4vLyB0b2RvXG5cbnZhciBmb28gPSBiYXI7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo1LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Miwibm8td2FybmluZy1jb21tZW50cyI6Mn0sImVudiI6e319fQ==).

### What did you expect to happen?

The sections

```
/*
 * todo
 */
```
and

```
/**
 * todo
 */
```
should be flagged for `no-warning-comments`.

### What actually happened?

Only

```
// todo
```

is flagged for `no-warning-comments`.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @Reeywhaar, thanks for the issue!

We were just discussing this in https://github.com/eslint/eslint/pull/16090, and it doesn't seem that the original issue https://github.com/eslint/eslint/issues/9884 was actually resolved at the time.

> ```
> /*
>  * todo
>  */
> ```
> 
> 
>     
>   
> 
> and
> 
> 
> ```
> /**
>  * todo
>  */
> ```
> 
> 
>     
>   
> 
> should be flagged for `no-warning-comments`.

I agree that this rule should treat these two comments as starting with 'todo' despite the extra `*` characters. But since the original issue was closed a long time ago, I would like more opinions from the team, especially because this change can produce more warnings.

**Comment by Reeywhaar** [comment]
I am not expert at `ast`, but is there a way to get "effective comment text" as [discussed previously](https://github.com/eslint/eslint/issues/9884#issuecomment-390371414)?

**Comment by lachlanhunt** [comment]
I had a go at stripping decorative punctuation in #16120 based on that previous discussion. The difficulty with this, though, is that there needs to be a clearly agreed upon definition of what counts as decorative punctuation in comments that should be stripped.

Rather than just stripping all leading and trailing punctuation, I looked for sequences of a few specific characters, and I picked characters that seemed like the most common to be encountered. Specifically, `-`, `*`, `=`, `/`, and `~`.

It's possible there are edge cases that I missed, and it's also possible there are other edge cases where it shouldn't strip punctuation, but it does.

The documentation for no-warning-comments should also get updated to explain the new functionality before merging it.

**Comment by mdjermanovic** [comment]
> Fixing this, however, would require adding the multiline (`m`) flag to the regex

I think this issue is only about comments that have "decoration" characters before the effective text starts. If we add the `m` flag, that would significantly change the meaning of `"location": "start"`. For example, the following comment would be detected as a comment that starts with `"todo"` term:

```js
/*
    this is not a
    todo comment
*/
```

**Comment by mdjermanovic** [comment]
A problem with removing some characters from the comment and then searching for terms in the stripped text is that terms in existing configurations can contain those characters.

For example, if we remove all leading `*`, then this would no longer match:

```js
/* eslint no-warning-comments: ["error", { terms: ["*TODO*"] } ] */

/*
    *TODO* add something
*/
```

A backward-compatible solution could be to keep the current behavior as is, and add an option to specify characters that should be skipped at the start of comments.

**Comment by lachlanhunt** [comment]
That was indeed my interpretation of the issue: Matching the start of any line in a comment, rather than just the start of the first line.

But implementing it the way you suggest would just require stripping decorative punctuation characters and leading whitespace up to the first non-punctuation character. It could be simplified further by only stripping asterisks and whitespace, if that's the desired behaviour. There could still be edge cases, like for example:

```js
{ terms: ["*TODO*"] }

/*
 *TODO* something
 */
```

Maybe instead of hardcoding the characters considered decorative, that could be configurable, with a sensible default setting.

```
"no-warning-comments": [1, {
    terms: ["TODO"],
    location: "start",
    commentDecorationChars: "*"
}]
```

**Comment by nzakas** [comment]
I believe the current behavior is intentional. I wouldn’t mind adding an option to support the JSDoc-style behavior, maybe allowing a user to specify the text to be skipped at the start of the comment?

I don’t think there’s a good heuristic for knowing which characters after the first asterisk should and shouldn’t be considered by this rule automatically.

**Comment by Reeywhaar** [comment]
> specify the text to be skipped at the start of the comment

As a workaround I've tried to amend rule as follows:
```
'no-warning-comments': ['warn', { terms: ['*\n  * TODO'], location: 'start' }]
```

And failed :-)

Also, I think I should mark that comment can be indented:
```
/*
 * todo:
 */

  /*
   * todo:
   */
```

So `text to be skipped` should be kind of regex.

**Comment by nzakas** [comment]
@eslint/eslint-tsc what do we think about adding a new option for the desired behavior?

**Comment by mdjermanovic** [comment]
I agree that this should be fixed by adding a new option.

The option could accept an array of characters to (optionally) skip at the start of the comment when checking whether it starts with disallowed terms? Similar to https://github.com/eslint/eslint/issues/16103#issuecomment-1180304234:

```js
"no-warning-comments": [1, {
    terms: ["TODO"],
    location: "start",
    commentDecorationChars: ["*"]
}]
```

These characters would be added to the [`prefix`](https://github.com/eslint/eslint/blob/1cdcbca8a961a057a9db40df412f249545befe2b/lib/rules/no-warning-comments.js#L88) part of the pattern (in the above example, `prefix` would be `"^[\\s*]*"`) so that they can also appear at the start of configured `terms`.

**Comment by nzakas** [comment]
I like it. 👍

**Comment by btmills** [comment]
@mdjermanovic's suggestion looks good!

**Comment by lachlanhunt** [comment]
I updated my PR to address the comments here. But I went with a single `decoration` property that takes a string of characters, rather than an array. It was easier to then just include that within the regex for location "start" using a character class.

given

```js
{ decoration: "*" }
```

The regex becomes `/^[\s\*]*ESCAPED_TERM\b/iu`

**Comment by mdjermanovic** [comment]
String as a list of characters might be confusing. For example, `"decoration": "/*"` (docs example from https://github.com/eslint/eslint/pull/16120) looks like it should skip `/*` sequences rather than individual characters `/` and `*`. I think arrays of characters like `["/", "*"]` in configs would be clearer. We can internally `.join("")` them and work with a string.

**Comment by lachlanhunt** [comment]
I thought about, but there's no easy way to enforce that each string in the array is a single character anyway. I could change it to allow either a string or array or strings, and then explain in the docs that they're equivalent.

**Comment by mdjermanovic** [comment]
We can use `pattern` to enforce single-character string values:

```js
commentDecorationChars: {
    type: "array",
    items: {
        type: "string",
        pattern: "^\\S$"
    },
    minItems: 1,
    uniqueItems: true
}
```

`\S` is to disallow specifying whitespace characters to emphasize that they're always included automatically.
